cloudinit/sources/helpers/vultr.py

```diff
@@ -0,0 +1,293 @@
+                "netmask": additional['netmask']
+            }
+            netcfg['subnets'].append(add)
+
+    # Check for additional IPv6's
+    additional_count = len(md['interfaces'][0]['ipv6']['additional'])
+    if "ipv6" in md['interfaces'][0] and additional_count > 0:
+        for additional in md['interfaces'][0]['ipv6']['additional']:
+            add = {
+                "type": "static6",
+                "control": "auto",
+                "address": additional['address'],
+                "netmask": additional['netmask']
+            }
+            netcfg['subnets'].append(add)
+
+    # Add config to template
+    return netcfg
+
+
+# Input Metadata and generate private network config part
+def generate_private_network_interface(md):
+    interface_name = get_interface_name(md['interfaces'][1]['mac'])
+    if not interface_name:
+        raise RuntimeError(
+            "Interface: %s could not be found on the system" %
+            md['interfaces'][1]['mac'])
+
+    netcfg = {
+        "name": interface_name,
+        "type": "physical",
+        "mac_address": md['interfaces'][1]['mac'],
+        "accept-ra": 1,
+        "subnets": [
+            {
+                "type": "static",
+                "control": "auto",
+                "address": md['interfaces'][1]['ipv4']['address'],
+                "netmask": md['interfaces'][1]['ipv4']['netmask']
+            }
+        ]
+    }
+
+    return netcfg
+
+
+# Generate the vendor config
+# This configuration is to replicate how
+# images are deployed on Vultr before Cloud-Init
+def generate_config(config):
+    LOG.debug("DS: %s", json.dumps(config))
+    md = get_cached_metadata(config)
+
+    # Grab the startup script
+    script = md['startup-script']
+
+    # Create vendor config
+    config_template = copy.deepcopy(md['vendor-config'])
+
+    # Add generated network parts
+    config_template['network'] = generate_network_config(config)
+
+    # Linux specific packages
+    if util.is_Linux():
+        config_template['packages'].append("ethtool")
+
+    # Define vendor script
+    vendor_script = "#!/bin/bash"
+
+    # Go through the interfaces
+    for netcfg in config_template['network']['config']:
+        # If the interface has a mac and is physical
+        if "mac_address" in netcfg and netcfg['type'] == "physical":
+            # Enable multi-queue on linux
+            # This is executed as a vendor script
+            if util.is_Linux():
+                # Set its multi-queue to num of cores as per RHEL Docs
+                name = netcfg['name']
+                command = "ethtool -L %s combined $(nproc --all)" % name
+                vendor_script = '%s\n%s' % (vendor_script, command)
+
+    # Add write_files if it is not present in the template
+    if 'write_files' not in config_template.keys():
+        config_template['write_files'] = []
+
+    # Add vendor script to config
+    config_template['write_files'].append(
```

Comment:
Is there any documentation for this? I can't seem to even find a reference for it.

Reply:
Sorry, but I'm going to have ask for verification on this. I couldn't find any examples in other datasources as well. I tried implementing it based on your first example and it seems util.py is trying to pass vendor_data_raw as a string to write_file even if it is a list.

```
2021-03-10 05:00:19,275 - util.py[WARNING]: failed stage init
2021-03-10 05:00:19,275 - util.py[DEBUG]: failed stage init
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/cmd/main.py", line 652, in status_wrapper
    ret = functor(name, args)
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/cmd/main.py", line 376, in main_init
    init.update()
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/stages.py", line 368, in update
    'vendordata')
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/stages.py", line 398, in _store_rawdata
    util.write_file(self._get_ipath('%s_raw' % datasource), data, 0o600)
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/util.py", line 1837, in write_file
    content = encode_text(content)
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/util.py", line 117, in encode_text
    return text.encode(encoding)
AttributeError: 'list' object has no attribute 'encode'`
```

```
    self.vendordata_raw = []
    for uscript in user_scripts:
        self.vendordata_raw.append(uscript)
    self.vendordata_raw.append("#cloud-config\n%s" % json.dumps(config))`
```

I've tested the package and the scripts are ran and the config is applied so it seems like this is intended to work.

Not sure if the example you provided is not quite right, this is a bug, or I simply misunderstood or did something wrong in my implementation. I went ahead and made an adjustment that checks for lists and converts it to a string and passes that in case this is indeed a bug. Seems to work with it. PR linked bellow, if this is the case and this is an appropriate fix. If it's not a bug or I did something wrong I can just close it and rewrite this portion another way, just need clarification on what I am doing wrong here if you don't mind.

https://github.com/canonical/cloud-init/pull/837